iPetCare/Application/Services/SpeciesService.cs

```diff
@@ -0,0 +1,183 @@
+
+            return new ServiceResponse<SpeciesGetAllDtoResponse>(HttpStatusCode.OK, dto);
+        }
+
+        public async Task<ServiceResponse<SpeciesGetDtoResponse>> GetAsync(int speciesId)
+        {
+            var currentUserName = _userAccessor.GetCurrentUsername();
+
+            if (currentUserName == null)
+                return new ServiceResponse<SpeciesGetDtoResponse>(HttpStatusCode.Unauthorized, "Brak uprawnień");
+
+            var currentUser = await _userManager.FindByNameAsync(currentUserName);
+            if (currentUser == null)
+                return new ServiceResponse<SpeciesGetDtoResponse>(HttpStatusCode.Unauthorized, "Brak uprawnień");
+
+            var species = await _context.Species.FindAsync(speciesId);
+
+            if (species == null)
+                return new ServiceResponse<SpeciesGetDtoResponse>(HttpStatusCode.BadRequest, "Nie istnieje taka rasa w bazie danych");
+
+            var dto = _mapper.Map<SpeciesGetDtoResponse>(species);
+
+            var races = await _context.Races.ToListAsync();
+
+            var filteredRaces = races.Where(race => race.SpeciesId == speciesId).ToList();
+
+            if (filteredRaces != null)
+                dto.Races = _mapper.Map<List<RaceDetailsGetDtoResponse>>(filteredRaces);
+
+            return new ServiceResponse<SpeciesGetDtoResponse>(HttpStatusCode.OK, dto);
+        }
+
+        public async Task<ServiceResponse<SpeciesUpdateDtoResponse>> UpdateAsync(int speciesId, SpeciesUpdateDtoRequest dto)
+        {
+            var currentUserName = _userAccessor.GetCurrentUsername();
+
+            if (currentUserName == null)
+                return new ServiceResponse<SpeciesUpdateDtoResponse>(HttpStatusCode.Unauthorized, "Brak uprawnień");
+
+            var currentUser = await _userManager.FindByNameAsync(currentUserName);
+            if (currentUser != null && currentUser.Role != Role.Administrator)
+                return new ServiceResponse<SpeciesUpdateDtoResponse>(HttpStatusCode.Forbidden, "Brak uprawnień");
+
+            if (await _context.Species.Where(x => x.Name == dto.Name).AnyAsync())
+                return new ServiceResponse<SpeciesUpdateDtoResponse>(HttpStatusCode.BadRequest, "Podana rasa już istnieje");
+
+            var species = _context.Species.Find(speciesId);
+
+            if (species == null)
+                return new ServiceResponse<SpeciesUpdateDtoResponse>(HttpStatusCode.BadRequest, "Nie istnieje taki gatunek w bazie danych");
+
+            species.Name = dto.Name;
+            int result = await _context.SaveChangesAsync();
+
+            if (result > 0)
+            {
+                var responseDto = _mapper.Map<SpeciesUpdateDtoResponse>(species);
+                return new ServiceResponse<SpeciesUpdateDtoResponse>(HttpStatusCode.OK, responseDto);
+
+            }
+
+            if (result == 0)
+                return new ServiceResponse<SpeciesUpdateDtoResponse>(HttpStatusCode.BadRequest, "Nie nastąpiła żadna zmiana");
```

Comment:
Ogólnie na brak zmiany zareagowałbym wcześniej i w inny sposób. Tzn sprawdziłbym czy nowa nazwa jest taka sama jak stara, i jak tak, to zwrócił Ok bez wysyłania żądania do bazy danych. Nie wiem czy powinniśmy traktowac brak zmiany jako błąd, w końcu nic złego się nie stało, wszystko się powiodło, tylko nazwa została podana taka sama jak jest obecnie. 🐈